Migrate Child Block Test to Playwright

pooja-muchandikar opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
## What?
<!-- In a few words, what is the PR actually doing? -->
Part of https://github.com/WordPress/gutenberg/issues/38851.
Migrate[ child-blocks.test.js ](https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-tests/specs/editor/plugins/child-blocks.test.js) to its Playwright version.
## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Part of https://github.com/WordPress/gutenberg/issues/38851.
## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
See[ MIGRATION.md](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/MIGRATION.md) for migration steps.
## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
Run` npm run test:e2e:playwright test/e2e/specs/editor/plugins/child-blocks.spec.js`

pooja-muchandikar committed:
Migrate Child Block Test to Playwright

pooja-muchandikar committed:
Fix Stylelint failing and modified the test case

pooja-muchandikar committed:
Fix failed CI

pooja-muchandikar commented:
Hi @kevin940726 @Mamaduka,

I hope you are doing great!

Could you help review this PR?

Thanks!

Mamaduka reviewed (commented):
Thank you, @pooja-muchandikar!

I think some of the locators can be improved. Especially ones that use classes instead of role-based selectors.

Recently, I did a similar migration (#51677); you can check the [`inner-blocks-allowed-blocks.spec.js`](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/specs/editor/plugins/inner-blocks-allowed-blocks.spec.js) file and use it as a blueprint.

pooja-muchandikar committed:
Update locators

pooja-muchandikar commented:
Hi @Mamaduka,

Thanks for sharing the feedbacks! All the feedbacks are addressed please review. 

Thanks!

Mamaduka commented:
Thanks, @pooja-muchandikar!

While changes look good, the two tests now check a different thing. They should test a block appender based on the original tests. Let's ensure that migrate tests are still checking for the same thing.

## Affected tests

* `shows up in a parent block`
* `display in a parent block with allowedItems`.


## Block appender
![CleanShot 2023-10-13 at 13 25 30](https://github.com/WordPress/gutenberg/assets/240569/0e6ae955-3dc9-4137-b68b-a99b0ac3dc87)

Mamaduka commented:
@pooja-muchandikar, let me know if you have any additional questions.

pooja-muchandikar committed:
Address Feedback

pooja-muchandikar commented:
Hi @Mamaduka, 

I have addressed the feedback and pushed the changes. Please review and let me know if there are any further suggestions. 

Thanks

Mamaduka commented on test/e2e/specs/editor/plugins/child-blocks.spec.js:
```diff
@@ -36,11 +36,9 @@ test.describe( 'Child Blocks', () => {
-			.click();
+		await page.click(
+			'[data-type="test/child-blocks-unrestricted-parent"] .block-editor-default-block-appender'
+		);
```
We should try and use [accessible selectors](https://github.com/WordPress/gutenberg/blob/a3f6e952db5c60e9197d0881b832d0efde186573/docs/contributors/code/e2e/README.md#use-accessible-selectors) when possible.

Example:

```js
await page
				.getByRole( 'document', {
					name: 'Block: Child Blocks Unrestricted Parent',
				} )
				.getByRole( 'button', {
					name: 'Add block',
				} )
				.click();
```

Mamaduka commented on test/e2e/specs/editor/plugins/child-blocks.spec.js:
```diff
@@ -63,11 +61,9 @@ test.describe( 'Child Blocks', () => {
-				name: 'Block: Child Blocks Restricted Parent',
-			} )
-			.click();
+		await page.click(
+			'[data-type="test/child-blocks-restricted-parent"] .block-editor-default-block-appender'
+		);
```
Same here.

pooja-muchandikar replied:
Hi @Mamaduka, I tried to use the solution here but it is not working as expected. 

`await page
				.getByRole( 'document', {
					name: 'Block: Child Blocks Restricted Parent',
				} )
				.getByRole( 'button', {
					name: 'Add block',
				} )
				.click();`

Mamaduka replied:
I think this should work now:

```js
await page
			.getByRole( 'document', {
				name: 'Block: Child Blocks Restricted Parent',
			} )
			.getByRole( 'button', {
				name: 'Add default block',
			} )
			.click();
```

pooja-muchandikar committed:
Address feedbacks

Mamaduka commented on test/e2e/specs/editor/plugins/child-blocks.spec.js:
```diff
@@ -0,0 +1,89 @@
+				name: 'Block: Child Blocks Unrestricted Parent',
+			} )
+			.getByRole( 'button', {
+				name: 'Add block',
```
```suggestion
				name: 'Add default block',
```

I just noticed that the original test was clicking on the default appender.

Mamaduka commented on test/e2e/specs/editor/plugins/child-blocks.spec.js:
```diff
@@ -0,0 +1,89 @@
+		} );
+
+		await blockInserter.click();
+		await expect( blockLibrary ).toBeVisible();
```
```suggestion
		await expect( blockLibrary ).toBeVisible();
		await expect( blockLibrary.getByRole( 'option' ) ).toContainText( [
			'Child Blocks Child',
		] );
```

This assertion was missing from the migration.

Mamaduka reviewed (commented):
I just ran the original test in headed mode and adjusted my suggestions. The tests should be passing now.

pooja-muchandikar committed:
Address feedbacks

pooja-muchandikar commented:
Hi @Mamaduka 

Feedbacks addressed.

Mamaduka commented on test/e2e/specs/editor/plugins/child-blocks.spec.js:
```diff
@@ -29,6 +29,9 @@ test.describe( 'Child Blocks', () => {
 		] );
+		expect(
+			await blockLibrary.getByRole( 'option' ).count()
+		).toBeGreaterThan( 10 );
```
I think this addition wasn't necessary here.

pooja-muchandikar replied:
Done

Mamaduka reviewed (approved):
Thank you, @pooja-muchandikar! One last nitpick and I think we're good to merge once CI checks are green.

pooja-muchandikar committed:
Update test case

pooja-muchandikar committed:
Update test case
